src/python/pants/backend/experimental/jvm/goals/coursier_integration_test.py

```diff
@@ -0,0 +1,199 @@
+# Copyright 2021 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import annotations
+
+import os
+import os.path
+from textwrap import dedent
+
+import pytest
+
+from pants.backend.experimental.jvm.goals.coursier import CoursierResolve
+from pants.backend.experimental.jvm.goals.coursier import rules as coursier_goal_rules
+from pants.backend.experimental.jvm.resolve.coursier_fetch import (
+    CoursierLockfileEntry,
+    CoursierResolvedLockfile,
+    MavenCoord,
+    MavenCoordinates,
+)
+from pants.backend.experimental.jvm.resolve.coursier_fetch import rules as coursier_fetch_rules
+from pants.backend.experimental.jvm.resolve.coursier_setup import rules as coursier_setup_rules
+from pants.backend.experimental.jvm.target_types import JvmDependencyLockfile
+from pants.backend.experimental.jvm.util_rules import rules as util_rules
+from pants.core.util_rules import config_files, source_files
+from pants.core.util_rules.external_tool import rules as external_tool_rules
+from pants.engine.fs import CreateDigest, Digest, FileContent, FileDigest
+from pants.testutil.rule_runner import RuleRunner
+
+
+@pytest.fixture
+def rule_runner() -> RuleRunner:
+    return RuleRunner(
+        preserve_tmpdirs=True,
+        rules=[
+            *config_files.rules(),
+            *coursier_fetch_rules(),
+            *coursier_goal_rules(),
+            *coursier_setup_rules(),
+            *external_tool_rules(),
+            *source_files.rules(),
+            *util_rules(),
+        ],
+        target_types=[JvmDependencyLockfile],
+    )
+
+
+def get_digest(rule_runner: RuleRunner, source_files: dict[str, str]) -> Digest:
+    files = [FileContent(path, content.encode()) for path, content in source_files.items()]
+    return rule_runner.request(Digest, [CreateDigest(files)])
+
+
+def test_coursier_resolve_creates_missing_lockfile(rule_runner: RuleRunner) -> None:
+    rule_runner.add_to_build_file(
```

Comment:
It's preferred to use the new `rule_runner.write_files({"BUILD": "...")}`. That was only added last week and it's not yet documented, and not all tests use it now. But it's easier to read and allows consolidating all the test setup into a single call.